Fix duplicate pattern error

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139321

vllm had an error when we were incorrectly stating two patterns are duplicates. See, comment inline:

For a particular generated pattern repr, store all the equivalent graphs that used to generate them. Because we ignore certain patterns in searching, but not in matching, use the graph to distinguish if two equivalent searches are actually different.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

eellison committed:
Fix incorrect duplicate pattern on patterns which have equivalent search patterns but different match

[ghstack-poisoned]

ProExpertProg commented:
Any reason `_seen_patterns` is not a property of `PatternMatcherPass`? Running into some weird lifetime semantics because a pattern can only be registered once.

eellison commented:
@ProExpertProg yes, agreed, they could be scoped to a pass.

eellison committed:
Update on "Fix duplicate pattern error"


vllm had an error when we were incorrectly stating two patterns are duplicates. See, comment inline:

For a particular generated pattern repr, store all the equivalent graphs that used to generate them. Because we ignore certain patterns in searching, but not in matching, use the graph to distinguish if two equivalent searches are actually different. Element are first inserted as graphs and lazily converted to their str reprs when there is a duplicate. If the graph is not present we will error on duplicate.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

eellison commented:
@pytorchbot rebase

eellison committed:
Update

[ghstack-poisoned]

shunting314 commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -1718,6 +1776,16 @@ def __init__(
+        # in searching, but not in matching, use the graph to distinguish if two equivalent
+        # searches are actually different.
+        # Element are first inserted as graphs and lazily converted to their str reprs when
+        # there is a duplicate. If the graph is not present we will error on duplicate
```
> If the graph is not present we will error on duplicate

You mean 'If the graph is ***already*** present we will error on duplicate' ?

shunting314 reviewed (commented):
So the PR change the duplication check from
- check string representation of pattern 
to
- check string representation of the fx graph generating the pattern
?

My questions is, why two patterns with the same string representation could be generated by different fx graphs (in the vllm case). Due to Ignored() pattern?

eellison committed:
Update

[ghstack-poisoned]

eellison commented:
@shunting314 yes, due to `Ignored`

eellison committed:
Update

[ghstack-poisoned]

shunting314 commented on test/inductor/test_pattern_matcher.py:
```diff
@@ -178,6 +178,66 @@ def fn2(a, b, c):
+            nonlocal count
+            count = patterns.apply(graph)
+
+        def custom_backend(
+            graph: torch.fx.GraphModule, example_inputs: Iterable[torch.Tensor]
+        ) -> Callable:
+            from torch._inductor import config
+
+            current_config = config.shallow_copy_dict()
+            from torch._inductor.compile_fx import compile_fx
+
+            current_config["post_grad_custom_post_pass"] = custom_pass
+            return compile_fx(graph, example_inputs, config_patches=current_config)
+
+        @torch.compile(backend=custom_backend)
```
Is this snippet equivalent to:
```
@torch.compile(options={"post_grad_custom_post_pass"=custom_pass})
```

ProExpertProg replied:
Today I learned...

eellison committed:
Update

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

malfet commented on test/inductor/test_pattern_matcher.py:
```diff
@@ -178,6 +178,66 @@ def fn2(a, b, c):
+            return x - 2
+
+        patterns = PatternMatcherPass()
+        torch.set_default_device("cuda")
```
Wouldn't that potentially affect other tests in a very weird ways? Shouldn't one better use `with torch.device("cuda"):` or just alloc next tensor on the device you want?

etaf replied:
Hi, @eellison can you help to fix the issue #143974 before re-land?  Just use GPU_TYPE instead of "cuda" as #143975 do.

eellison replied:
@malfet was reusing repro and forgot to update, agreed. although having said that - there are a few tests which use this pattern.. so if it is harmful, we should fix them.

jeanschmidt commented:
@pytorchbot revert -m "breaking internal signals" -c ghfirst

eellison committed:
Update

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

clee2000 commented:
If this gets reverted in OSS, please also revert D67865938 internally.  It was modified during diff train, giving it some internal changes
